packages/graphback-datasync/src/providers/DataSyncConflictProvider.ts

```diff
@@ -0,0 +1,96 @@
+import { ModelDefinition, GraphbackContext, TransformType, FieldTransform, NoDataError } from '@graphback/core';
+import { DataSyncModelConflictConfig, serverSideConflictResolution, DataSyncFieldNames, ConflictResolutionStrategy, ConflictStateMap, ConflictError } from '../util';
+import { MongoDeltaHelper } from '../deltaHelper';
+import { DataSyncMongoDBDataProvider } from './DatasyncMongoDBDataProvider';
+
+type ConflictHandler = (conflict: ConflictStateMap, deltaHelper: MongoDeltaHelper) => any
+
+const throwOnConflictHandler: ConflictHandler = (conflict: ConflictStateMap) => {
+  throw new ConflictError(conflict);
+}
+
+/**
+ * Da
+ */
+export class DataSyncConflictMongoDBDataProvider<Type = any> extends DataSyncMongoDBDataProvider<Type> {
+  protected conflictConfig: DataSyncModelConflictConfig
+  protected deltaHelper: MongoDeltaHelper
+  protected handleUpdateConflicts: ConflictHandler
+
+  public constructor(model: ModelDefinition, client: any, dataSyncConflictConfig: DataSyncModelConflictConfig) {
+    super(model, client);
+    this.conflictConfig = dataSyncConflictConfig;
+
+    this.handleUpdateConflicts = dataSyncConflictConfig.conflictResolution?.resolveUpdate || throwOnConflictHandler;
+
+    if (this.conflictConfig.conflictResolution === serverSideConflictResolution) {
+      this.deltaHelper = new MongoDeltaHelper(model, client);
+    }
+  }
+
+  public async create(data: any, context: GraphbackContext): Promise<Type> {
+    const contextWithoutSelectedFields: GraphbackContext = { ...context }
+    contextWithoutSelectedFields.graphback.options.selectedFields = [];
+
+    if (this.conflictConfig.enabled) {
+      data[DataSyncFieldNames.version] = 1;
+    }
+
+    const result = await super.create(data, contextWithoutSelectedFields);
+
+    await this.deltaHelper?.insertDiff(result).catch((e: any) => { console.error(`Error in inserting delta: ${e}`) });
+
+
+    return result;
+  }
+
+  public async update(updateDocument: any, context: GraphbackContext): Promise<Type> {
+    const { _id } = updateDocument;
+
+    const MAX_RETRIES = 3;
+    const updateFilter = { _id, [DataSyncFieldNames.deleted]: false };
+
+    for(let i = 0; i < MAX_RETRIES; i++) {
+      const serverStateData = await this.db.collection(this.collectionName).findOne({ _id, [DataSyncFieldNames.deleted]: false });
+
+      if (serverStateData){
+
+        
+        if (updateDocument[DataSyncFieldNames.version] !== serverStateData[DataSyncFieldNames.version]) {
+          const conflict: ConflictStateMap = {
```

Comment:
This doesn't mean conflict. 
It means only that versions are different. 
We need diffs to know if conflict occured.

Server data might have different fields than client as well.